### ISSUE TITLE
No matching version found for ndarray@1.1.0

### ISSUE THREAD

**Issue by dmbarry86** [plotly/plotly.js#5925]
When reviewing our weekly Dependabot updates I came across an issue with our migration from Plotly 2.4.1 to 2.5.0. We now get the error shown below when trying to install dependencies. The issue was shown in our CI pipeline which runs `npm ci`. 

![image](https://user-images.githubusercontent.com/8976377/132201204-6125a996-d233-423b-93d7-56e82886eccb.png)

I've attached our package files from 2.4.1 which work fine and the updated ones from 2.5.0 which break. You can run `npm ci` with the 2.5.0 files to reproduce.

[error.zip](https://github.com/plotly/plotly.js/files/7115145/error.zip)

**Comment by archmoj** [comment]
What is your npm version?

**Comment by dmbarry86** [comment]
> What is your npm version?

Locally it's 6.14.12. Our CI pipeline (GitHub Actions) would also be 6.x as we run on Node 14.x which I believe comes with 6.x npm as default.

**Comment by archmoj** [comment]
Upgrading to npm7 and node16 could help fix the warning. See https://github.com/plotly/plotly.js/pull/5922 and https://github.com/plotly/plotly.js/pull/5921#issuecomment-912610011.

**Comment by dmbarry86** [comment]
Do you no longer support node 14 then? Note that we get an error not a warning.

**Comment by archmoj** [comment]
node12 and node14 still worked for us with npm ci. The was a warning with npm i which upgrading to npm7 fixed it.
What is your package.json look like?

**Comment by dmbarry86** [comment]
> What is your package.json look like?

I attached this above to the original report.

**Comment by archmoj** [comment]
Could you please try using ploly.js-dist instead?

**Comment by dmbarry86** [comment]
> Could you please try using ploly.js-dist instead?

This appears to fix the issue. Are there any downsides to switching over to this package?

Are you able to investigate what is causing the original problem? I see some of our other dependencies make use of the original ndarray rather than the plotly forked version. Do you think this could be the cause?

**Comment by archmoj** [comment]
> > Could you please try using ploly.js-dist instead?
> 
> This appears to fix the issue. Are there any downsides to switching over to this package?
> 

There shouldn't be any downside. It should also download and build faster. In addition you may try the minified package plotly.js-dist-min too.

**Comment by archmoj** [comment]
> > 
> Are you able to investigate what is causing the original problem? I see some of our other dependencies make use of the original ndarray rather than the plotly forked version. Do you think this could be the cause?

Which other package require ndarray?
Could you try npm i ndarray --save ?

**Comment by dmbarry86** [comment]
FYI, switching to Node 16.x and NPM 7.x solves the error. We're not quite ready to switch to Node 16 yet in our project but we will keep in mind. For now maybe we just stay with Plotly 2.4.1 until we are ready to migrate.

**Comment by dmbarry86** [comment]
> Which other package require ndarray?

Apologies, I was wrong there. I was manually inspecting the package-lock.json and thought another dependency relied on it. It appears to just be plotly dependencies.

Here's the output from `npm ls ndarray` in our working 2.4.1
![image](https://user-images.githubusercontent.com/8976377/132224753-5e3a7647-bf3f-45a2-9e49-f8c1f00774b4.png)


Here's the output from `npm ls ndarray` in our broken 2.5.0
![image](https://user-images.githubusercontent.com/8976377/132224728-7fd95c49-992b-486e-ba55-b61c46c17bd2.png)

**Comment by archmoj** [comment]
npm uninstall plotly.js 
npm install plotly.js-dist --save

Should fix the issue for you.

**Comment by dmbarry86** [comment]
If I uninstall plotly.js and then reinstall plot.js I now get a new error when running `npm ci`. The new error this time is "No matching version found for 3d-view@2.1.0.". This is still with npm 6.x by the way as we have shown that 7.x already works fine.

**Comment by archmoj** [comment]
Not reinstalling plotly.js, 

npm install plotly.js-dist --save

is installing ploltly.js-dist

**Comment by dmbarry86** [comment]
> Not reinstalling plotly.js,
> 
> npm install plotly.js-dist --save
> 
> is installing ploltly.js-dist

Yes I'm aware. We've already proved above that plotly.js-dist works just fine. I'm trying to help debug the issue with plotly.js and `npm ci`.

**Comment by archmoj** [comment]
> > Not reinstalling plotly.js,
> > npm install plotly.js-dist --save
> > is installing ploltly.js-dist
> 
> Yes I'm aware. We've already proved above that plotly.js-dist works just fine. I'm trying to help debug the issue with plotly.js and `npm ci`.

What's the reasoning behind not using ploly.js-dist  then?

**Comment by dmbarry86** [comment]
> What's the reasoning behind not using ploly.js-dist then?

We're actually using react-plotly in our React project. Switching from plotly.js to plotly.js-dist resolves the `npm ci` issue but we then get errors when importing from react-plotly.js. Is there a way to use react-plotly with plotly.js-dist?

![image](https://user-images.githubusercontent.com/8976377/132231096-d9c3e7ff-46ce-4a7b-9ff1-cce2431e24fd.png)

**Comment by archmoj** [comment]
> > What's the reasoning behind not using ploly.js-dist then?
> 
> We're actually using react-plotly in our React project. Switching from plotly.js to plotly.js-dist resolves the `npm ci` issue but we then get errors when importing from react-plotly.js. Is there a way to use react-plotly with plotly.js-dist?
> 
> ![image](https://user-images.githubusercontent.com/8976377/132231096-d9c3e7ff-46ce-4a7b-9ff1-cce2431e24fd.png)

I see. Please open an issue or pull request on react-plotly.js and paste the link here.

**Comment by dmbarry86** [comment]
Do I take it from your comments so far that you don't plan to try and debug and fix the issue with plotly.js 2.5.0, npm 6.x and `npm ci`? If that's the case then we just stick with version 2.4.1 for now.

**Comment by archmoj** [comment]
> Do I take it from your comments so far that you don't plan to try and debug and fix the issue with plotly.js 2.5.0, npm 6.x and `npm ci`? If that's the case then we just stick with version 2.4.1 for now.

Yes please you could continue using 2.4.1 (or bump to 2.4.2) until the react-plotly.js issue addressed or as discussed upgrade npm to v7.

**Comment by dmbarry86** [comment]
FWIW I can reproduce the error in a completely fresh project that has just plotly.js version 2.5.0 installed. Personally I would consider this a breaking change (or a bug) as you now no longer support npm 6.x and npm ci.

How to reproduce:
* Run on npm 6.x (the default with node 14)
* `npm init newProject`
* `npm install plotly.js --save`
* `npm ci`
* observe error regaridng no matching version found for xxx dependency

![image](https://user-images.githubusercontent.com/8976377/132234162-028b1a3c-b5a3-47ba-b86f-32e794d5ee90.png)

**Comment by archmoj** [comment]
> FWIW I can reproduce the error in a completely fresh project that has just plotly.js version 2.5.0 installed. Personally I would consider this a breaking change (or a bug) as you now no longer support npm 6.x and npm ci.
> 
> How to reproduce:
> 
> * Run on npm 6.x (the default with node 14)
> * `npm init newProject`
> * `npm install plotly.js --save`
> * `npm ci`
> * observe error regarding no matching version found for xxx dependency
> 
> ![image](https://user-images.githubusercontent.com/8976377/132234162-028b1a3c-b5a3-47ba-b86f-32e794d5ee90.png)

Since v2 importing ploly.js-dist is the preferred and documented way of requiring, this may not be considered as a breaking change IMHO.

**Comment by dmbarry86** [comment]
> this may not be considered as a breaking change IMHO.

Speaking as somebody who now has broken code, I beg to differ. The documentation for react-plotly clearly states to still install plotly.js. We clearly disagree here and that's fine. We can just stick to version 2.4.1 / 2.4.2 for our use case.

**Comment by alexcjohnson** [comment]
@archmoj I understand your point but I think we should still make an effort to figure out what's going on here. If it's just a matter of npm6 not recognizing the link format we're using it'll be a simple fix. If it requires us to give all these forked packages different names (`@plotly/ndarray` etc) then that may be a bigger lift but perhaps a cleaner representation of what we're doing anyway.

BTW one thing I notice, not sure if this has anything to do with the error, but the `repository` field in our `ndarray` fork [points to the original repo](https://github.com/plotly/ndarray/blob/70ec67233be65de444307a6916aba2655e28bd84/package.json#L23), not our own.

**Comment by archmoj** [comment]
@dmbarry86 FYI - I opened https://github.com/plotly/react-plotly.js/pull/260.
Any chance you could try the commit with `npm=6`?

**Comment by archmoj** [comment]
> @dmbarry86 FYI - I opened [plotly/react-plotly.js#260](https://github.com/plotly/react-plotly.js/pull/260).
> Any chance you could try the commit with `npm=6`?

by installing
```json
"react-plotly.js": "git://github.com/plotly/react-plotly.js#40dfac18bd98d64b013bd726f5c212aef0d9e4d1"
```
in `package.json`.

**Comment by dmbarry86** [comment]
> @dmbarry86 FYI - I opened [plotly/react-plotly.js#260](https://github.com/plotly/react-plotly.js/pull/260).
> Any chance you could try the commit with `npm=6`?

Doesn't seem to be working for me. I get an error in our components that use react-plotly.js
![image](https://user-images.githubusercontent.com/8976377/132390821-f1aa22ae-5ba9-461e-8912-3fa9d6b6ef6f.png)

Package.json is now as follows
![image](https://user-images.githubusercontent.com/8976377/132390881-ac4cee3d-f834-4143-8459-8c9f003963d9.png)

I checked node_modules folder and it has successfully pulled down the version from your branch so I'm not sure exactly what is wrong here.

**Comment by dmbarry86** [comment]
> I checked node_modules folder and it has successfully pulled down the version from your branch so I'm not sure exactly what is wrong here.

Oh actually it has only pulled down the root project files. There isn't actually any source code in the folder.
![image](https://user-images.githubusercontent.com/8976377/132391582-a31366f4-1d61-4c5a-9608-68d748dd70b8.png)

**Comment by archmoj** [comment]
@dmbarry86 thanks for testing.
Wondering what happens if `plotly.js-dist` is moved to dependencies like https://github.com/plotly/react-plotly.js/commit/4ecac06b1af7152f6cca7d1d0c577b9c594c9b9b

Could you possibly try 
```json
"react-plotly.js": "git://github.com/plotly/react-plotly.js#4ecac06b1af7152f6cca7d1d0c577b9c594c9b9b"
```
as well?

**Comment by dmbarry86** [comment]
@archmoj still the same problem with no source code inside node_modules\react-plotly.js folder.

**Comment by dmbarry86** [comment]
@archmoj I wonder if the problem is because the npmignore file in the repository is ignoring src. folder. Maybe you could temporarily remove this and create new commit so we can test.

**Comment by archmoj** [comment]
Unfortunately this week we may not be able to allocate more time investigating this problem. 
But please let us know if you found a solution for this namely at the `react-plotly.js` level.

**Comment by dmbarry86** [comment]
> But please let us know if you found a solution for this

See my comment about regarding npmignore. I think for me to be able to npm install from GitHub and test your PR you will need to (temporarily) remove src from npmignore.

**Comment by dmbarry86** [comment]
@archmoj I've forked my own version of your branch and removed src from npmignore. I can now confirm that your PR is working as expected with npm version 6.

**Comment by dmbarry86** [comment]
This is a good article FWIW. It recommends the use of files in package.json rather than npmignore. https://medium.com/@jdxcode/for-the-love-of-god-dont-use-npmignore-f93c08909d8d

**Comment by archmoj** [comment]
@dmbarry86 Fantastic, please open a PR https://github.com/plotly/react-plotly.js/compare/switch-to-plotly.js-dist...dmbarry86:switch-to-plotly.js-dist

**Comment by dmbarry86** [comment]
@archmoj It doesn't need a PR from me. That removal of src was just temporary to allow local testing. Your own PR is the one that should be merged.

**Comment by archmoj** [comment]
> @archmoj I've forked my own version of your branch and removed src from npmignore. I can now confirm that your PR is working as expected with npm version 6.

Thanks for testing. Please leave a comment on this pull https://github.com/plotly/react-plotly.js/pull/260.
